torch/fx/experimental/symbolic_shapes.py

```diff
@@ -10,7 +11,36 @@
 
 aten = torch.ops.aten
 
-__all__ = ["has_symbolic_sizes_strides", "create_contiguous", "is_symbolic_op", "handle_symbolic_op", "PySymInt", "ShapeEnv"]
+__all__ = [
+    "has_symbolic_sizes_strides", "create_contiguous", "is_symbolic_op", "handle_symbolic_op", "PySymInt", "ShapeEnv",
+    "SymDispatchMode"
+]
+
+SYM_FUNCTION_MODE = None
+
+# We don't bother with the metaclass as all of the dispatching logic happens
+# entirely from Python
+#
+# Didn't bother with ancestors for now, unlikely to have multiple modes for
+# symints right now
+
+class SymDispatchMode:
+    def __sym_dispatch__(self, func, types, args, kwargs):
+        raise NotImplementedError()
+
+    def __enter__(self):
+        global SYM_FUNCTION_MODE
+        old = SYM_FUNCTION_MODE
+        if hasattr(self, "inner"):
+            raise RuntimeError(f"{self} has already been used as a mode. Please use a fresh version")
+        else:
+            self.inner = old
```

Comment:
why do we nest them... i would have expected it all to be under one environment / set of shape guards

Reply:
The most likely nesting situation is for `get_isolated_graph`, i.e. you are retracing an FX graph and you want to speculatively trace a decomposition and see if you like it or not before taking it. You would have two SymInt proxy modes in that case.